src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp

```diff
@@ -32,23 +33,19 @@ std::unique_ptr<Intent> Tactic::getNextIntent()
         return std::unique_ptr<Intent>{};
     }
 
-    if (intent_sequence)
+    auto next_intent = getNextIntentHelper();
+    if (done_ && loop_forever)
     {
-        // Calculate and return the next Intent
-        intent_sequence();
-        auto next_intent = intent_sequence.get();
-        // The tactic is considered done once a nullptr is returned
-        // We set the done_ variable here in addition to below (once the coroutine is
-        // done) to make sure that the done() function will always return true at the same
-        // time a nullptr is returned (which also indicates the Tactic is done)
-        done_ = !static_cast<bool>(next_intent);
-        return next_intent;
+        // If the tactic is done and is supposed to loop forever, we re-create the
+        // intent_sequence which "restarts" the coroutine. We then run the coroutine
+        // again, and take the new result so that we don't return a nullptr while
+        // restarting the Tactic
```

Comment:
Would be good to note that the first thing a restarted coroutine returns is a `std::nullptr` here, just to be super clear

Reply:
That behavior is documented with the `calculateNextIntentWrapper` function and isn't the relevant part here. I'll try clarify in this comment that if we detect a `nullptr` is returned (indicating the Tactic is done), we restart the coroutine and take the new value to replace the `nullptr` so that any callers don't think the Tactic is done.

Reply:
In fact I could probably just not mention the nullptr stuff because we're basing this behavior on `done_`, not the return value. It's technically not the responsibility of this code to know about the returned nullptr, just to restart the coroutine when it's done.